### ISSUE TITLE
make Rowan traits aware

### ISSUE THREAD

**Issue by dalehenrich** [GemTalk/Rowan#783]
[Magritte](https://github.com/magritte-metamodel/magritte), a project that has been ported to GLASS for the last decade (see [Magritte3](https://github.com/GsDevKit/Magritte3) and [Magritte2](http://seaside.gemtalksystems.com/ss/Magritte2.html)), is currently using Pharo traits and there is [demand to be able to use the latest incarnation of Magritte in GemStone](https://github.com/magritte-metamodel/magritte/issues/257). The [T2F2T project](https://github.com/GsDevKit/T2F2T-Conversion) meets the immediate need to be able to translate the package format from tonel to filetree, [assuming that manual traits expansion has been performed in Pharo](https://github.com/magritte-metamodel/magritte/issues/257#issuecomment-1007610206), but things get a bit more complicated when translating the packages from filetree to tonel, since the extra flattened definitions should be ignored.

It should be possible to teach Rowan about traits definitions and automatically:
1. do the expansion when translating from tonel to filetree (straightforward)
2. ignore the expanded definitions when translating from filetree to tonel (more complicated)

Of course in the long term it is desirable to find a way to support traits natively in GemStone so that the need for flattening, but that is a discussion for another day

**Comment by AllenOtis** [comment]
The use case for GemStone indexing subsystem is ,   a trait file  is a class.st  file containing  a  class defintion (declared as a trait) and method definitions .   If a trait exists as an actual class,   instance creation is not allowed, and it contains no methods.  It could contain a class variable which is a dictionary of selectors and method source strings .
For a  class that uses a trait each method definition in the trait file is compiled in that class .  instance variable and class variable references in method definitions in the trait file are resolved separately for each class into which the trait is included

**Comment by dalehenrich** [comment]
@AllenOtis, here's an example of [Traits in a tonel file](https://github.com/GemTalk/JadeiteForPharo/blob/main/Jadeite-Traits/JadeiteToolbarTrait.trait.st):
```
Trait {
	#name : 'JadeiteToolbarTrait',
	#instVars : [
		'toolbar'
	],
	#category : 'Jadeite-Traits'
}

{ #category : 'toolbar support' }
JadeiteToolbarTrait classSide >> toolbarCommands [

	^ { 
		  JadeiteAbortCommand.
		  JadeiteCommitCommand }
]

{ #category : 'initialization' }
JadeiteToolbarTrait >> addToolbarCommandsTo: aGroup [

	self class toolbarCommands do: [ :each | 
		aGroup register: (each forSpecContext: self) ]
]

{ #category : 'actions' }
JadeiteToolbarTrait >> clientCommitIdDescription [

	| ws | 
	ws := WriteStream on: String new. 
	self clientCommitIdDescriptionOn: ws. 
	^ws contents
]

{ #category : 'actions' }
JadeiteToolbarTrait >> clientCommitIdDescriptionOn: ws [

	self jadeiteClientGitRepositories doWithIndex: [ :repoName :index |
		| repo |
		repo := IceRepository registry detect: [ :ea | ea name = repoName ].
		index = 1 ifFalse:[ws cr]. 
		ws
			nextPutAll: repoName;
			space;
			nextPutAll: repo properties commitish shortId ]
]

{ #category : 'actions' }
JadeiteToolbarTrait >> displayClientCommitId: workspace [

	| commitIdDescription |
	commitIdDescription := self clientCommitIdDescription.
	^ workspace text: commitIdDescription
]

{ #category : 'actions' }
JadeiteToolbarTrait >> displayClientandServerCommitIds: workspace [

	| ws |
	ws := WriteStream on: String new.
	ws
		nextPutAll: self clientCommitIdDescription;
		cr;
		nextPutAll: '---------------------------------------------';
		cr;
		nextPutAll: self serverCommitIdDescription.
	^ workspace text: ws contents
]

{ #category : 'initialization' }
JadeiteToolbarTrait >> initializeToolbar [

	self newJadeiteToolbar.
	self layout
		add: #toolbar
		expand: false
		fill: false
		padding: 0
]

{ #category : 'actions' }
JadeiteToolbarTrait >> jadeiteClientGitRepositories [

	^ #( 'JadeiteForPharo' 'PharoGemStoneFFI' 'RemoteServiceReplication' )
]

{ #category : 'actions' }
JadeiteToolbarTrait >> jadeiteServerGitRepositories [

	^ #( 'RowanClientServices' )
]

{ #category : 'initialization' }
JadeiteToolbarTrait >> newJadeiteToolbar [

	toolbar := self newToolbar.
	toolbar
		addStyle: 'stToolbar';
		fillWith: self toolbarActions.
	toolbar
		addItem: (SpToolbarButtonPresenter new
				 label: 'Workspace';
				 icon: (self iconNamed: #workspace);
				 help: 'opens GemStone workspace';
				 action: [ self openWorkspace ];
				 yourself);
		addItem: (SpToolbarButtonPresenter new
				 label: 'Browser';
				 icon: (self iconNamed: #smallSystemBrowser);
				 help: 'open Browser';
				 action: [ self openBrowser ];
				 yourself);
		addItem: (SpToolbarButtonPresenter new
				 label: 'SUnit Browser';
				 icon: (self iconNamed: #smallPaint);
				 help: 'open SUnit Browser and Test Runner';
				 action: [ self openSUnitBrowser ];
				 yourself);
		yourself.
	^toolbar
]

{ #category : 'actions' }
JadeiteToolbarTrait >> openBrowser [

	| inst | 
	inst := JadeiteBrowser new owner: self. 
	inst
		application: self application;
		gciSession: self gciSession.
	inst open.
	inst addStyle: 'jadeWorkspace'.
	^inst
]

{ #category : 'actions' }
JadeiteToolbarTrait >> openCommitIdWindow [

	| workspace windowCaption |
	windowCaption := 'Jadeite Git Commit Ids'.
	self connection
		ifNil: [
			workspace := JadeiteNonSessionApplication new start.
			self displayClientCommitId: workspace ]
		ifNotNil: [
			workspace := JadeiteWorkspaceApplication showOnSession: self connection.
			self displayClientandServerCommitIds: workspace ].
	workspace setCaption: windowCaption.
	workspace hasUnacceptedEdits: false.
	^workspace
]

{ #category : 'actions' }
JadeiteToolbarTrait >> openSUnitBrowser [

	| inst |
	inst := JadeiteSUnitPresenter basicNew.
	inst
		application: self application;
		gciSession: self gciSession;
		initialize.
	inst open.
	inst addStyle: 'jadeWorkspace'
]

{ #category : 'actions' }
JadeiteToolbarTrait >> openWorkspace [

	JadeiteWorkspaceApplication showOnSession: self gciSession library
]

{ #category : 'actions' }
JadeiteToolbarTrait >> serverCommitIdDescription [

	| ws | 
	ws := WriteStream on: String new. 
	self serverCommitIdDescriptionOn: ws. 
	^ws contents
]

{ #category : 'actions' }
JadeiteToolbarTrait >> serverCommitIdDescriptionOn: ws [

	| browserService |
	browserService := RowanBrowserService new.
	browserService command: #updateProjects.
	RowanServiceUpdater current
		issueCommand: browserService
		session: (GciSession new library: self connection) onCompletion: [ :unused |
	self jadeiteServerGitRepositories do: [ :projectName |
		| project |
		project := browserService projects detect: [ :projectService |
			           projectService name = projectName ].
		ws
			nextPutAll: projectName;
			space;
			nextPutAll: project sha;
			cr ]]
]

{ #category : 'initialization' }
JadeiteToolbarTrait >> toolbarActions [

	^ CmCommandGroup forSpec
		in: [ :this | self addToolbarCommandsTo: this ];
		yourself
]
```
[Traits declaration in a tonel file](https://github.com/GemTalk/JadeiteForPharo/blob/main/Jadeite-Browser/JadeiteBrowser.class.st) uses:# message in creation declares which Traits are to be added to a class:
```
Class {
	#name : 'JadeiteBrowser',
	#superclass : 'JadeitePresenterWithAutoCommitButton',
	#traits : 'JadeiteMethodsTrait + JadeiteProjectsTrait + JadeiteToolbarTrait + JadeiteClassTrait + JadeiteFilterTrait + JadeitePackageTrait',
	#classTraits : 'JadeiteMethodsTrait classTrait + JadeiteProjectsTrait classTrait + JadeiteToolbarTrait classTrait + JadeiteClassTrait classTrait + JadeiteFilterTrait classTrait + JadeitePackageTrait classTrait',
	#instVars : [
		'tabs',
		'toolbar'
	],
	#category : 'Jadeite-Browser'
}
```
So I think that on disk and possibly in behavior, we should be able to match Pharo pretty closely ...

**Comment by dalehenrich** [comment]
will be available in 3.7.2
